src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -82,63 +92,361 @@
+     *
+     * @param structureName name of the structure to load
+     * @return the input stream or null
+     */
+    public static InputStream getStream(final String structureName)
+    {
+        final Structures.StructureName sn = new Structures.StructureName(structureName);
+        InputStream inputstream;
+        if (Structures.SCHEMATICS_CACHE.equals(sn.getPrefix()))
+        {
+            return Structure.getStreamFromFolder(Structure.getCachedSchematicsFolder(), structureName);
+        }
+        else if (Structures.SCHEMATICS_SCAN.equals(sn.getPrefix()))
+        {
+            return Structure.getStreamFromFolder(Structure.getClientSchematicsFolder(), structureName);
+        }
+        else if (!Structures.SCHEMATICS_PREFIX.equals(sn.getPrefix()))
+        {
+            return null;
+        }
+        else
+        {
+            //Look in the folder first
+            inputstream = Structure.getStreamFromFolder(Structure.getSchematicsFolder(), structureName);
+            if (inputstream == null && !Configurations.ignoreSchematicsFromJar)
+            {
+                inputstream = Structure.getStreamFromJar(structureName);
+            }
+        }
+
+        if (inputstream == null)
+        {
+            Log.getLogger().warn("Structure: Couldn't find any structure with this name " + structureName);
+        }
+
+        return inputstream;
+    }
+
+    /**
+     * get a input stream for a schematic within a specif folder.
+     *
+     * @param folder        where to load it from.
+     * @param structureName name of the structure to load.
+     * @return the input stream or null
+     */
+    private static InputStream getStreamFromFolder(final File folder, final String structureName)
+    {
+        try
+        {
+            if (folder.exists())
+            {
+                //We need to check that we stay within the correct folder
+                final File nbtFile = new File(folder.getPath() + "/" + structureName + ".nbt");
+                if (nbtFile.toURI().normalize().getPath().startsWith(folder.toURI().normalize().getPath()))
                 {
-                    inputstream = new FileInputStream(decorationFolder.getPath() + "/" + structureName + ".nbt");
+                    return new FileInputStream(folder.getPath() + "/" + structureName + ".nbt");
                 }
                 else
                 {
-                    throw new FileNotFoundException("Unable to find structure: " + structureName);
+                    Log.getLogger().error("Structure: Illegal structure name \"" + structureName + "\"");
                 }
             }
-            catch (final FileNotFoundException e)
+            else
             {
-                Log.getLogger().warn("Couldn't find any structure with this name anywhere", e);
+                throw new FileNotFoundException("Unable to find structure: " + structureName);
             }
         }
+        catch (final FileNotFoundException e)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)